### ISSUE TITLE
Deserializing error: Unable to instantiate class

### ISSUE THREAD

**Issue by RaulUrtecho** [k-paxian/dart-json-mapper#227]
Hi nice package, I hope you can assist me. 
I have the following models:

```dart
@jsonSerializable
class Filters extends Equatable {
  final String? author;
  final List<String>? colDisplay;
  final List<String>? colOrder;
  final String? created;
  final Criteria? criteria;
  @JsonProperty(name: "default")
  final bool? isDefault;
  final String? lastModified;
  final bool? public;
  final List<String>? tags;
  final String? title;
  @JsonProperty(ignoreIfNull: true)
  final String? uuid;

  Filters({
    this.author,
    this.colDisplay = const [],
    this.colOrder = const [],
    this.created,
    this.criteria,
    this.isDefault,
    this.lastModified,
    this.public,
    this.tags,
    this.title,
    this.uuid,
  });

  @override
  @JsonProperty(ignore: true)
  List<Object?> get props =>
      [
        author,
        colDisplay,
        colOrder,
        created,
        criteria,
        isDefault,
        lastModified,
        public,
        tags,
        title,
        uuid,
      ];

  @override
  @JsonProperty(ignore: true)
  bool? get stringify => super.stringify;

  Filters copyWith({
    String? author,
    List<String>? colDisplay,
    List<String>? colOrder,
    String? created,
    Criteria? criteria,
    bool? isDefault,
    String? lastModified,
    bool? public,
    List<String>? tags,
    String? title,
    String? uuid,
  }) {
    return Filters(
      author: author ?? this.author,
      colDisplay: colDisplay ?? this.colDisplay,
      colOrder: colOrder ?? this.colOrder,
      created: created ?? this.created,
      criteria: criteria ?? this.criteria,
      isDefault: isDefault ?? this.isDefault,
      lastModified: lastModified ?? '${DateTime.now().toUtc().asBackendDate}Z',
      public: public ?? this.public,
      tags: tags ?? this.tags,
      title: title ?? this.title,
      uuid: uuid ?? this.uuid,
    );
  }

  static Filters empty({
    required String title,
    required bool public,
    required bool isDefault,
  }) => Filters(title: title, public: public, isDefault: isDefault);
}

@jsonSerializable
class Criteria {
  final String? sortBy;
  final String? sortDirection;
  final List<Sort>? sorts;
  final int? rows;
  @JsonProperty(name: "q")
  final String? query;
  @JsonProperty(name: "fqs")
  final List<Filter> filters;

  Criteria({this.sortBy, this.sortDirection, this.sorts, this.rows, this.query, this.filters = const []});

  Criteria copyWith({
    String? sortBy,
    String? sortDirection,
    List<Sort>? sorts,
    int? rows,
    String? query,
    List<Filter>? filters,
  }) {
    return Criteria(
      sortBy: sortBy ?? this.sortBy,
      sortDirection: sortDirection ?? this.sortDirection,
      sorts: sorts ?? this.sorts,
      rows: rows ?? this.rows,
      query: query ?? this.query,
      filters: filters ?? this.filters,
    );
  }
}

@jsonSerializable
class Sort {
  final String? field;
  final String? direction;

  Sort(this.field, this.direction);

  @override
  @JsonProperty(ignore: true)
  List<Object?> get props => [field, direction];
}

@jsonSerializable
class Filter extends Equatable{
  final String? field;
  final String? operator;
  final dynamic value;

  Filter(this.field, this.operator, this.value);

  @override
  @JsonProperty(ignore: true)
  List<Object?> get props => [field, value];

  @override
  @JsonProperty(ignore: true)
  bool? get stringify => super.stringify;

  @JsonProperty(ignore: true)
  bool get not => operator?.startsWith('n') == true;

  @JsonProperty(ignore: true)
  String get filterOperation {
    return '';
  }
}
```
When running my app on Android and deserializing I encounter the following error:

`(Unable to instantiate class 'Criteria' with null named arguments [Symbol("sortBy"), Symbol("sortDirection")])`

I have already tried several things in the model criteria such as: changing from mutable to immutable, positional instead of named, default values ​​in properties and constructs, using non-nullables, and nothing has worked for me, except when I removed the renamed fields
```dart
  @JsonProperty(name: "q")
  final String? query;
  @JsonProperty(name: "fqs")
  final List<Filter> filters;
```
 which is very strange.
 
 I would appreciate any insights you can offer as to whether this error is on my end or if something was overlooked.

**Comment by k-paxian** [comment]
Can you also show me the JSON input you are trying to deserialize so it would be easier to create a unit test out of your case

**Comment by RaulUrtecho** [comment]
@k-paxian Sorry for the delay here the json:
```
{
  "author": "raul",
  "colDisplay": [
    "a",
    "b",
    "c"
  ],
  "colOrder": [
    "a",
    "b",
    "c"
  ],
  "created": "2024-09-19T20:23:12.543Z",
  "criteria": {
    "sortBy": null,
    "sortDirection": null,
    "sorts": [
      {
        "field": "abc",
        "direction": "asc"
      }
    ],
    "rows": 10,
    "q": {
      "field": "_text_",
      "operator": "abc",
      "value": "abc"
    },
    "fqs": [
       {
           "field": "a",
           "operator": "b",
           "value": [
             "a",
             "b"
           ]
         },
    ]
  },
  "default": false,
  "lastModified": "2024-09-19T20:23:12.543Z",
  "public": false,
  "tags": [
    "a"
  ],
  "title": "abc",
  "uuid": "123"
}
```

**Comment by k-paxian** [comment]
Ok there are two issues

* One is that you are trying to map an object 
```json
"q": {
      "field": "_text_",
      "operator": "abc",
      "value": "abc"
    },
```
to  a String field:

```dart
  @JsonProperty(name: "q")
  final String? query;
```

Obviously you need to have a class there to describe the shape of the Query. That's why it's not possible to instantiate your `Criteria` class

* Other issue is that you are getting a misleading error text
`(Unable to instantiate class 'Criteria' with null named arguments [Symbol("sortBy"), Symbol("sortDirection")])`
That's because only one kind of an error is handled by the mapper - when you define the non nullable field and is going to assign null value to the field - then class instantiation also fails.

So all in all this issue can be rephrased: "Incorrect error message is shown for your case"  😃

**Comment by RaulUrtecho** [comment]
@k-paxian Yes, I changed the type to `Map`, and the error was resolved. 

Thanks for clarifying the message, honestly, it confused me at first and led me to do more testing in other parts, overlooking the obvious solution. A dev should know when it’s time for a break to clear their mind and come back with fresh perspective, haha.

Thanks again for the help and pointing that out!

**Comment by k-paxian** [comment]
Fixed in `2.2.14`
